---
title: "SPER-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SPER-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup}
library(SPER)
##  Other necessary libraries
library(Seurat)
# library(dplyr)
```

# Example: adult mouse brain spatial transciptomics dataset

## Dataset
You can download all the datasets through Dropbox: [link provided]
- Spatial transcriptomics (ST) data are located at "data/Mouse_Brain_ST_data.h5", which should be loaded though Seurat 
  - It contains the transcriptomics information of all spots in the provided mouse brain sample. One can treat each spot as one RNA-seq outcome that has thousands of genes.
- Distance matrix is located at "data/rounded_distance_matrix.txt"
  - This big matrix records the distance information between every two spots. 
  - One can get this matrix from the physical coordinates just by few steps. This means that you can actually obtain it from the ST dataset.
- Cell-type spatial compositional data are located at "data/spatial_data.txt"
  - It contains the spatial distributions of all cell types with the help of distance matrix.
  - One can either use the 'raw' data calculated from cell-type decomposition algorithms, or use the data after CoKriging, which can model the desired inter-cell-type spatial relationship and also remove background noise 
  - Here we provide the after-CoKriging compositional data
  
```{r}
####  Spatial transcriptomics data
head(ST_expression_data)

####  Coordinates & Distance matrix
head(ST_coordinates)
spot_dist <- as.matrix(dist(ST_coordinates))
round_dist_mat <- ceiling(spot_dist / 120) * 120
round_dist_mat[which(round_dist_mat > 1800)] <- 1800

####  spatial compositional data
head(spatial_comp_data)
```

## Example: 
- With given data, we can run SPER and get the outcome as a list. The list has each gene as the name of an element where it actually contains the spatial paired expression matrix between the gene and all cell types. 

```{r}
res <- SPER(dist.mat = round_dist_mat,
            dist.list = (0:15)*120,
            ST.mat = ST_expression_data,
            CoDa.data = spatial_comp_data)
res[["L2.3.IT"]][1:10, 1:10]
```

- Above results shows the SPER between L2/3 IT cell type and all genes. For example, for the 1st column, the values represents the spatial association between *Grp* and L2/3 IT cells at the given distance (0, 120 microns, 240 microns, etc.). Higher value means the gene is spatially more associated with the cells. 

- However, it will be impractical to compare the whole column between pairs. Therefore, we can apply a customized weight on the column so that we can integrate the values into one score, which is SPER score.

```{r}
# custom_weights <- c(rep(1/5, 5), rep(0, 11))
custom_weights <- c(0, 0, rep(1/3, 3), rep(0, 11))
SPER_res <- weightSPER(res,
                       custom_weights)
head(SPER_res)
```

- 

## Example of visualizaiton: gene spatial distribution
```{r}
gg_gene("Grp",
        ST_expression_data,
        coord = ST_coordinates,
        title.anno = "Gene")
```

